File: backend/sdtm_logger.py
import json
import uuid
from datetime import datetime, timezone
from pathlib import Path

LOG_DIR = Path("logs")
LOG_DIR.mkdir(exist_ok=True)


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def log_exposure(study_id: str, subject_id: str, treatment: str, ndc: str):
    """
    SDTM EX Domain — Exposure event.
    Records verified pill ingestion with timestamp.
    """
    record = {
        "STUDYID": study_id,
        "DOMAIN": "EX",
        "USUBJID": subject_id,
        "EXSEQ": str(uuid.uuid4()),
        "EXTRT": treatment,
        "EXNDC": ndc,
        "EXSTDTC": _now(),
        "EXSTAT": "VERIFIED_BY_CNN",
        "EXMETHOD": "WEBCAM_CNN_CLASSIFICATION"
    }
    _append_log("EX", subject_id, record)


def log_questionnaire(study_id: str, subject_id: str, qstest: str, qsorres: str):
    """
    SDTM QS Domain — Questionnaire response.
    Records a single SMAQ question answer.
    """
    record = {
        "STUDYID": study_id,
        "DOMAIN": "QS",
        "USUBJID": subject_id,
        "QSSEQ": str(uuid.uuid4()),
        "QSCAT": "SMAQ",
        "QSTEST": qstest,
        "QSORRES": qsorres,
        "QSDTC": _now()
    }
    _append_log("QS", subject_id, record)


def _append_log(domain: str, subject_id: str, record: dict):
    log_file = LOG_DIR / f"{domain}_{subject_id}.jsonl"
    with open(log_file, "a") as f:
        f.write(json.dumps(record) + "\n")

File: backend/sdtm_logger.js
const fs   = require('fs');
const path = require('path');
const crypto = require('crypto');

const LOG_DIR = path.join(__dirname, 'logs');
if (!fs.existsSync(LOG_DIR)) fs.mkdirSync(LOG_DIR);

/**
 * Writes a single SDTM-compatible record to a JSONL audit log.
 *
 * EX domain: dose_verified events
 * QS domain: diary_submitted events (questionnaire responses)
 */
function logAdherenceEvent(payload) {
  const { subject_id, event, timestamp, domain } = payload;

  let record;

  if (event === 'dose_verified' || event === 'prevented_dosing_error') {
    // SDTM EX Domain — Exposure
    record = {
      STUDYID:  process.env.TRIAL_ID || 'TRIAL-001',
      DOMAIN:   'EX',
      USUBJID:  subject_id,
      EXSEQ:    crypto.randomUUID(),
      EXEVENT:  event,
      EXSTAT:   event === 'dose_verified' ? 'VERIFIED_BY_EDGE_CNN' : 'PREVENTED_DOSING_ERROR',
      EXCONF:   payload.confidence ?? null,
      EXMETHOD: 'TFJS_MOBILENETV2_EDGE_INFERENCE',
      EXSTDTC:  timestamp
    };
  } else if (event === 'diary_submitted') {
    // SDTM QS Domain — Questionnaire
    record = {
      STUDYID: process.env.TRIAL_ID || 'TRIAL-001',
      DOMAIN:  'QS',
      USUBJID: subject_id,
      QSSEQ:   crypto.randomUUID(),
      QSCAT:   'SMAQ',
      QSORRES: payload.responses,
      QSDTC:   timestamp
    };
  } else {
    record = { STUDYID: 'TRIAL-001', DOMAIN: 'AUDIT', USUBJID: subject_id, ...payload };
  }

  const domainKey = record.DOMAIN || 'AUDIT';
  const logFile   = path.join(LOG_DIR, `${domainKey}_${subject_id}.jsonl`);
  fs.appendFileSync(logFile, JSON.stringify(record) + '\n');
}

module.exports = { logAdherenceEvent };
